core/src/main/scala/org/http4s/HeaderKey.scala

```diff
@@ -15,6 +17,8 @@ sealed trait HeaderKey {
   final def unapply(header: Header): Option[HeaderT] = matchHeader(header)
 
   override def toString: String = s"HeaderKey($name})"
+
+  def fromString(s: String): ParseResult[HeaderT]
```

Comment:
We already had a `fromString` on `Content-Length`.  I'm not sure how often we'll call `.fromString` on most headers, but completing the codec and filling in the arbitraries does let scalacheck do the heavy lifting in testing our parsers.

I considered a type provider macro to replace the reflection and also provide this method for what is currently solved by `Header.Internal`, but my macro-fu is not yet strong enough.